pymc3/variational/opvi.py

```diff
@@ -0,0 +1,522 @@
+
+    def log_q_W_local(self, z):
+        """
+        log_q_W samples over q for local vars
+        Gradient wrt mu, rho in density parametrization
+        is set to zero to lower variance of ELBO
+        """
+        if not self.local_vars:
+            return 0
+        logp = []
+        for var in self.local_vars:
+            mu = self.known[var][0].ravel()
+            rho = self.known[var][1].ravel()
+            mu = self.scale_grad(mu)
+            rho = self.scale_grad(rho)
+            x = self.view(z, var.name, reshape=False)
+            q = log_normal(x, mu, rho=rho)
+            logp.append(q.sum() * var.scaling)
+        return tt.sum(logp)
+
+    def log_q_W_global(self, z):
+        """
+        log_q_W samples over q for global vars
+        """
+        raise NotImplementedError
+
+    def logq(self, z):
+        """
+        Total logq for approximation
+        """
+        return self.log_q_W_global(z) + self.log_q_W_local(z)
+
+    def view(self, space, name, reshape=True):
+        """
+        Construct view on a variable from flattened `space`
+
+        Parameters
+        ----------
+        space : space to take view of variable from
+        name : str
+            name of variable
+        reshape : bool
+            whether to reshape variable from vectorized view
+
+        Returns
+        -------
+        variable view
+        """
+
+        slc = self._view[name].slc
+        _, _, _shape, dtype = self._view[name]
+        if space.ndim == 2:
+            view = space[:, slc]
+        elif space.ndim < 2:
+            view = space[slc]
+        else:
+            raise ValueError('Space should have no more than 2 dims, got %d' % space.ndim)
+        if reshape:
+            if len(_shape) > 0:
+                if isinstance(space, tt.TensorVariable):
+                    shape = tt.concatenate([space.shape[:-1],
+                                            tt.as_tensor(_shape)])
+                else:
+                    shape = np.concatenate([space.shape[:-1],
+                                            _shape])
+
+            else:
+                shape = space.shape[:-1]
+            view = view.reshape(shape)
+        return view.astype(dtype)
+
+    @property
+    def total_size(self):
+        return self.order.dimensions
+
+    @property
+    def local_size(self):
+        size = sum([0] + [v.dsize for v in self.local_vars])
+        return size
+
+    @property
+    def global_size(self):
+        return self.total_size - self.local_size
+
+    @property
+    def local_slc(self):
+        return slice(0, self.local_size)
+
+    @property
+    def global_slc(self):
+        return slice(self.local_size, self.total_size)
+
+
+class KL(Operator):
+    def apply(self, f):
+        """KL divergence between posterior and approximation for input `z`
+            :math:`z ~ Approximation`
+        """
+        z = self.input
+        return self.logq(z) - self.logp(z)
+
+
+class LS(Operator):
```

Comment:
Yes, this does not work yet. I have great vacations and will resume to work in 14 days:)
Welcome for feedback)